Hylafaxplus

Yarny0 opened the pull request:
This pull requests contains three commits

- add myself to the maintainers list (is that ok?)
- add a package hylafaxplus
- add a nixos module to configure the hylafax server service (including secondary tasks such as spool directory maintainance and cleanup)

###### Motivation for this change

Send and receive faxes with NixOS.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS: 18.03 & unstable
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`): NixOS 18.03
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

There is another Pull Request to that end: https://github.com/NixOS/nixpkgs/pull/19392 , involved parties: @RamKromberg , @Mic92

Yarny0 committed:
maintainers: add 'yarny' (= myself)

Yarny0 committed:
HylaFAX+: init at 5.6.0

Create the top-level packages attribute
'hylafaxplus' that builds HylaFAX+ .

Note:
The nobody uid and the nogroup gid
are hardcoded in the package.
The package build recipe file
contains options to modify these ids.

Yarny0 committed:
modules: HylaFAX server configuration

This commit adds the following
* the uucp user
* options for HylaFAX server to control startup and modems
* systemd services for HylaFAX server processes
  including faxgettys for modems
* systemd services to maintain the HylaFAX spool area,
  including cleanup with faxcron and faxqclean
* default configuration for all server processes
  for a minimal working configuration

Some notes:

* HylaFAX configuration cannot be initialized with faxsetup
  (as it would be common on other Linux distributions).
  The hylafaxplus package contains a template spool area.
* Modems are controlled by faxgetty.
  Send-only configuration (modems controlled by faxq)
  is not supported by this configuration setup.
* To enable the service, one or more modems must be defined with
  config.services.hylafax.modems .
* Sending mail *should* work:
  HylaFAX will use whatever is in
  config.services.mail.sendmailSetuidWrapper.program
  unless overridden with the sendmailPath option.
* The admin has to create a hosts.hfaxd file somewhere
  (e.g. in /etc) before enabling HylaFAX.
  This file controls access to the server (see hosts.hfaxd(5) ).
  Sadly, HylaFAX does not permit account-based access
  control as is accepts connections via TCP only.
* Active fax polling should work; I can't test it.
* Passive fax polling is not supported by HylaFAX.
* Pager transmissions (with sendpage) are disabled by default.
  I have never tested or used these.
* Incoming data/voice/"extern"al calls
  won't be handled by default.
  I have never tested or used these.

Mic92 commented on nixos/modules/services/networking/hylafax/options.nix:
```diff
@@ -0,0 +1,364 @@
+
+    package = mkOption {
+      type = package;
+      default = pkgs.hylafaxplus;
```
What would be the alternative package to `hylafaxplus`? Can you add an example?

Yarny0 replied:
I thought it would be a good idea to leave room for HylaFAX (not +). I faintly recall that a couple of years ago I had to use the not-plus version as HylaFAX+ did some things that broke my workflows. However, I don't recall the exact reason.

Would it be better to remove this option? I could do that as I currently need the plus-version only, and others might use the overlays-mechanism of nixpkgs.

Mic92 commented on nixos/modules/services/networking/hylafax/options.nix:
```diff
@@ -0,0 +1,364 @@
+
+    faxcron.enable.spoolInit = mkEnableOption ''
+      Purge old files from the spooling area with
+      `faxcron`
```
The docbook equivalent, would be <literal>faxcron</literal>

Mic92 commented on nixos/modules/services/networking/hylafax/options.nix:
```diff
@@ -0,0 +1,364 @@
+      example = "daily";
+      description = ''
+        Purge old files from the spooling area with
+        `faxcron` with the given frequency
```
here too.

Mic92 commented on nixos/modules/services/networking/hylafax/options.nix:
```diff
@@ -0,0 +1,364 @@
+      example = "daily";
+      description = ''
+        Purge old files from the spooling area with
+        `faxcron` with the given frequency
```
here too.

Mic92 commented on nixos/modules/services/networking/hylafax/options.nix:
```diff
@@ -0,0 +1,364 @@
+        Enable or suppress job archiving:
+        `false` (the default!) disables archiving completely,
+        `true` forces archiving of all jobs,
+        while `null` only enables archiving for jobs
```
here too.

Mic92 commented on nixos/modules/services/networking/hylafax/systemd.nix:
```diff
@@ -0,0 +1,263 @@
+      ''${cfg.package}/spool/bin/faxqclean''
+      ''-q "${cfg.spoolAreaPath}"''
+      ''-v''
+      (optionalString (cfg.faxqclean.archiving!=false) ''-a'')
```
nitpick: `optionalString (cfg.faxqclean.archiving) ''-a''` is a bit more idiomatic and common.

Yarny0 replied:
Not really in this case: `archiving` can also be set to `null`, which means that it will archive only jobs that are flagged for archiving. I fear that `optionalString` will misunderstand `null` as boolean value.

Mic92 replied:
Could you make it a `enum` instead of boolean option? `faxqclean.archiving = null;` is not so self-explaining.

Mic92 commented on nixos/modules/services/networking/hylafax/systemd.nix:
```diff
@@ -0,0 +1,263 @@
+      ''-q "${cfg.spoolAreaPath}"''
+      ''-v''
+      (optionalString (cfg.faxqclean.archiving!=false) ''-a'')
+      (optionalString (cfg.faxqclean.archiving==true)  ''-A'')
```
see above

Mic92 commented on pkgs/servers/hylafaxplus/default.nix:
```diff
@@ -0,0 +1,95 @@
+  # for a couple of standard binaries in the `PATH` and
+  # hardcode their absolute paths in the new package.
+  buildInputs = [
+    coreutils  # for `base64` command
```
you can assume `coreutils` to be always present.

Mic92 commented on pkgs/servers/hylafaxplus/default.nix:
```diff
@@ -0,0 +1,95 @@
+    openldap  # optional
+    pam  # optional
+  ];
+  postPatch = ''. "${postPatch}"'';
```
you can assume `postPatch` to to contain spaces, since the nix store does not.

Yarny0 replied:
[I guess that's a "not"]
Is that certain? I recall reading somewhere that the directory of the nix store may be changed (at the price of a complete rebuild of all packages). I generally try to be on the safe side with spaces or other shell characters.

Mic92 replied:
No, spaces in the nix path are not supported. We prefer simpler code instead.

Mic92 commented:
When enabling hylafax (`services.hylafax.enable = true`) the service fails to start:

```
● hylafax-spool.service - HylaFAX spool area preparation
   Loaded: loaded (/nix/store/pcxnbf9qdn9w0fqrisfybpihy1c5z31z-unit-hylafax-spool.service/hylafax-spool.service; linked; vendor preset: enabled)
   Active: failed (Result: exit-code) since Fri 2018-08-17 14:32:05 BST; 4min 17s ago
     Docs: man:hylafax-server(4)
  Process: 1142 ExecStart=/nix/store/8x88gkskijziagzkxsvil5jgnl49qpd1-unit-script-hylafax-spool-start (code=exited, status=1/FAILURE)
 Main PID: 1142 (code=exited, status=1/FAILURE)

Aug 17 14:32:05 turingmachine systemd[1]: Starting HylaFAX spool area preparation...
Aug 17 14:32:05 turingmachine 8x88gkskijziagzkxsvil5jgnl49qpd1-unit-script-hylafax-spool-start[1142]: cp: cannot stat '/nix/store/f69a1iijx3nir4c08pcdzvq9aqdfsd6j-hylafax-config-modems/*': No such file or directory
Aug 17 14:32:05 turingmachine systemd[1]: hylafax-spool.service: Main process exited, code=exited, status=1/FAILURE
Aug 17 14:32:05 turingmachine systemd[1]: hylafax-spool.service: Failed with result 'exit-code'.
Aug 17 14:32:05 turingmachine systemd[1]: Failed to start HylaFAX spool area preparation.

```

Yarny0 commented:
> When enabling hylafax (`services.hylafax.enable = true`) the service fails to start

Yes, one needs to define at least one modem. I'm wondering if it would be useful to permit starting HylaFAX without modems, but there is probably no use to this. With the latest version of the pull request, I added an assertion that complains about missing modems. Additionally, I added a note on the need to define modems in the description of the `config.services.hylafax.modems` option.

There is another option with a similar effect: `config.services.hylafax.userAccessFile` points to `/etc/hosts.hfaxd` by default. This file must exist and be accessible by the `uucp` user only. The admin may create it with a proper `config.environment.etc."hosts.hfaxd"` configuration. I added a corresponding note in the option description. I'm not sure if I can/should check this more thoroughly.

Finally, I changed the names of some supporting build scripts of the package to be more consistent.

Mic92 committed:
nixos/hylafax: show correct option in warning message.
